index.bs

```diff
@@ -677,14 +692,515 @@ id=] must be the same as the [=window handle=].
+  value: null,
+}
+
+NullObject = {
+  type: "null",
+  value: null,
+}
+
+StringObject = {
+  type: "string",
+  value: text,
+}
+
+SpecialNumber = "NaN" / "-0" / "+Infinity" / "-Infinity";
+
+NumberObject = {
+  type: "number",
+  value: number / SpecialNumber,
+}
+
+BooleanObject = {
+  type: "boolean",
+  value: bool,
+}
+
+BigIntObject = {
+  type: "bigint",
+  value: text,
+}
+
+SymbolObject = {
+  type: "symbol",
+  objectId: ObjectId,
+}
+
+ArrayObject = {
+  type: "array",
+  objectId: ObjectId,
+  value?: [RemoteObject]
+}
+
+ObjectObject = {
+  type: "object",
+  objectId: ObjectId,
+  value?: {*text: RemoteObject}
+}
+
+FunctionObject = {
+  type: "function",
+  objectId: ObjectId,
+}
+
+RegexObject = {
+  type: "regex",
+  objectId: ObjectId,
+  value: text
+}
+
+DateObject = {
+  type: "date",
+  objectId: ObjectId,
+  value: text
+}
+
+MapObject = {
+  type: "map",
+  objectId: ObjectId,
+}
+
+SetObject = {
+  type: "set",
+  objectId: ObjectId
+}
+
+WeakMapObject = {
+  type: "map",
+  objectId: ObjectId,
+}
+
+WeakSetObject = {
+  type: "set",
+  objectId: ObjectId,
+}
+
+IteratorObject = {
+  type: "iterator",
+  objectId: ObjectId,
+}
+
+GeneratorObject = {
+  type: "iterator",
+  objectId: ObjectId,
+}
+
+ErrorObject = {
+  type: "error",
+  objectId: ObjectId,
+}
+
+ProxyObject = {
+  type: "proxy",
+  objectId: ObjectId,
+}
+
+PromiseObject = {
+  type: "promise",
+  objectId: ObjectId,
+}
+
+TypedArrayObject = {
+  type: "typedarray",
+  objectId: ObjectId,
+}
+
+ArrayBufferObject = {
+  type: "arraybuffer",
+  objectId: ObjectId,
+}
+
+NodeObject = {
+  type: "node",
+  objectId: ObjectId,
+  value?: NodeValue,
+}
+
+NodeValue = {
+  nodeType: uint,
+  nodeName: text,
+  localName: text,
+  nodeValue: text,
+  childNodeCount: uint,
+  children?: [*NodeObject],
+  attributes?: {*text: text},
+}
+
+WindowProxyObject = {
+  type: "window",
+  objectId: ObjectId,
+}
+
+
+```
+
+Issue: Add WASM types?
+
+Issue: NodeValue needs more attributes
+
+Issue: Should WindowProxy get attributes in a similar style to Node?
+
+<div algorithm>
+
+To <dfn>serialize as a remote object</dfn> given an |object|, a |max depth|,
+|node details|, and a |set of known objects|:
+
+  1. Switch on the type of |object|:
+
+    <dl>
+      <dt>Undefined
+      <dd>Let |remote object| be an [=object=] matching the `UndefinedObject`
+      production in the [=local end definition=].
+
+      <dt>Null
+      <dd>Let |remote object| be an [=object=] matching the `NullObject`
+      production in the [=local end definition=].
+
+      <dt>String
+      <dd>Let |remote object| be an [=object=] matching the `StringObject`
+      production in the [=local end definition=], with the <code>value</code>
+      property set to |object|.
+
+      <dt>Number
+      <dd>
+      1. Switch on the value of |object|:
+        <dl>
+          <dt>NaN
+          <dt>Let |value| be <code>"NaN"</code>
+          <dt>-0
+          <dt>Let |value| be <code>"-0"</code>
```

Comment:
Hmm, tricky. This seems like it'll screw things up for people who write their code not expecting any strings, while not being able to represent it is also problematic...

Lazy out: what do other protocols do?

Reply:
CDP has "value" and "unserializableValue" fields which are used in cases like this.